From 036e7c0340c406daaa7223d893f010555b0f0cc9 Mon Sep 17 00:00:00 2001
From: Vignesh Goutham Ganesh <vgg@amazon.com>
Date: Tue, 6 Dec 2022 15:42:02 -0600
Subject: [PATCH 04/19] RHEL support and improvements

- Fix redhat 9 cloud-init feature flags bug
- Fix ensure iptables present for rhel
- disable gpg for extra rpms
- Remove old kernels
- Add dracut cmd to generate initramfs with all drivers for rhel raw
- Add proxy, register with satellite, and pull packages from satellite support to redhat subscription manager
- Fix state for Ansible RHSM repository module

Signed-off-by: Vignesh Goutham Ganesh <vgg@amazon.com>
---
 images/capi/ansible/roles/node/tasks/main.yml |  3 +
 .../capi/ansible/roles/node/tasks/redhat.yml  | 32 +++++++
 .../capi/ansible/roles/setup/tasks/redhat.yml | 83 +++++++++++++++++++
 3 files changed, 118 insertions(+)
 create mode 100644 images/capi/ansible/roles/node/tasks/redhat.yml

diff --git a/images/capi/ansible/roles/node/tasks/main.yml b/images/capi/ansible/roles/node/tasks/main.yml
index 5a596aa9f..17f886a99 100644
--- a/images/capi/ansible/roles/node/tasks/main.yml
+++ b/images/capi/ansible/roles/node/tasks/main.yml
@@ -18,6 +18,9 @@
 - ansible.builtin.import_tasks: amazonLinux.yml
   when: ansible_distribution == "Amazon"
 
+- import_tasks: redhat.yml
+  when: ansible_distribution == "RedHat"
+
 # This is required until https://github.com/ansible/ansible/issues/77537 is fixed and used.
 - name: Override Flatcar's OS family
   ansible.builtin.set_fact:
diff --git a/images/capi/ansible/roles/node/tasks/redhat.yml b/images/capi/ansible/roles/node/tasks/redhat.yml
new file mode 100644
index 000000000..b2133b6f1
--- /dev/null
+++ b/images/capi/ansible/roles/node/tasks/redhat.yml
@@ -0,0 +1,32 @@
+# Copyright 2020 The Kubernetes Authors.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+# http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+---
+- name: Get the list of installed kernels
+  ansible.builtin.shell: "rpm -q kernel --qf '%{version}-%{release}\n' | sort -V"
+  register: installed_kernels
+
+- name: Get the current running kernel
+  ansible.builtin.command: "uname -r"
+  register: current_kernel
+
+- name: Filter out the current kernels
+  set_fact:
+    old_kernels: "{{ installed_kernels.stdout_lines | difference([current_kernel.stdout]) |list }}"
+
+- name: Remove old kernels
+  ansible.builtin.yum:
+    name: "kernel-{{ item }}"
+    state: absent
+  loop: "{{ old_kernels }}"
\ No newline at end of file
diff --git a/images/capi/ansible/roles/setup/tasks/redhat.yml b/images/capi/ansible/roles/setup/tasks/redhat.yml
index 030e7f5cf..d9240c4df 100644
--- a/images/capi/ansible/roles/setup/tasks/redhat.yml
+++ b/images/capi/ansible/roles/setup/tasks/redhat.yml
@@ -22,6 +22,74 @@
     - ansible_distribution == "RedHat"
     - lookup('env', 'RHSM_USER') | length > 0
     - lookup('env', 'RHSM_PASS') | length > 0
+    - rhsm_server_hostname == ""
+    - rhsm_server_proxy_hostname == ""
+
+- name: RHEL subscription with proxy
+  redhat_subscription:
+    state: present
+    username: "{{ lookup('env', 'RHSM_USER') }}"
+    password: "{{ lookup('env', 'RHSM_PASS') }}"
+    auto_attach: true
+    server_proxy_hostname: "{{ rhsm_server_proxy_hostname | default(omit) }}"
+    server_proxy_port: "{{ rhsm_server_proxy_port | default(omit) }}"
+  when:
+    - ansible_distribution == "RedHat"
+    - lookup('env', 'RHSM_USER') | length > 0
+    - lookup('env', 'RHSM_PASS') | length > 0
+    - rhsm_server_hostname == ""
+    - rhsm_server_proxy_hostname != ""
+
+- name: Download Katello CA Consumer RPM from Satellite Server
+  get_url:
+    url: "http://{{ rhsm_server_hostname }}/pub/katello-ca-consumer-latest.noarch.rpm"
+    dest: /tmp/katello.rpm
+    mode: 0755
+    owner: root
+    group: root
+  when:
+    - ansible_distribution == "RedHat"
+    - rhsm_server_hostname != ""
+
+- name: Install Katello CA Consumer RPM from Satellite Server
+  yum:
+    name: /tmp/katello.rpm
+    state: present
+    disable_gpg_check: true
+    lock_timeout: 60
+  when:
+    - ansible_distribution == "RedHat"
+    - rhsm_server_hostname != ""
+
+- name: RHEL subscription with satellite using activation key
+  redhat_subscription:
+    state: present
+    activationkey: "{{ lookup('env', 'RHSM_ACTIVATION_KEY') }}"
+    org_id: "{{ lookup('env', 'RHSM_ORG_ID') }}"
+    server_hostname: "{{ rhsm_server_hostname | default(omit) }}"
+    release: "{{ rhsm_server_release_version }}"
+  when:
+    - ansible_distribution == "RedHat"
+    - lookup('env', 'RHSM_ACTIVATION_KEY') | length > 0
+    - lookup('env', 'RHSM_ORG_ID') | length > 0
+    - rhsm_server_hostname != ""
+    - rhsm_server_proxy_hostname == ""
+
+- name: RHEL subscription with satellite and proxy
+  redhat_subscription:
+    state: present
+    activationkey: "{{ lookup('env', 'RHSM_ACTIVATION_KEY') }}"
+    org_id: "{{ lookup('env', 'RHSM_ORG_ID') }}"
+    server_hostname: "{{ rhsm_server_hostname | default(omit) }}"
+    server_proxy_hostname: "{{ rhsm_server_proxy_hostname | default(omit) }}"
+    server_proxy_port: "{{ rhsm_server_proxy_port | default(omit) }}"
+    release: "{{ rhsm_server_release_version }}"
+  when:
+    - ansible_distribution == "RedHat"
+    - lookup('env', 'RHSM_ACTIVATION_KEY') | length > 0
+    - lookup('env', 'RHSM_ORG_ID') | length > 0
+    - rhsm_server_hostname != ""
+    - rhsm_server_proxy_hostname != ""
 
 - name: Perform dnf clean
   ansible.builtin.command: /usr/bin/yum -y clean all
@@ -70,3 +138,18 @@
     name: "{{ extra_rpms.split() }}"
     state: present
     lock_timeout: 60
+    disable_gpg_check: true
+
+- name: ensure iptables present for redhat 8
+  yum:
+    name: "iptables"
+    state: present
+    lock_timeout: 60
+  when: ansible_os_family == "RedHat" and ansible_distribution == "RedHat" and ansible_distribution_major_version == "8"
+
+- name: ensure iptables present for redhat 9
+  yum:
+    name: "iptables-nft"
+    state: present
+    lock_timeout: 60
+  when: ansible_os_family == "RedHat" and ansible_distribution == "RedHat" and ansible_distribution_major_version == "9"
-- 
2.45.2
